src/System.Collections.NonGeneric/tests/Hashtable/GetEnumeratorTests.cs

```diff
@@ -0,0 +1,244 @@
+            Assert.Throws<InvalidOperationException>(() =>
+                             {
+                                 strRetKey = (string)dicEnum1.Key;
+                             }
+            );
+
+            //Entry dicEnum1 = (IDictionaryEnumerator *)hash1.GetEnumerator();
+            dicEnum1 = hash1.GetEnumerator();
+            while (dicEnum1.MoveNext() != false)
+            {
+                ;//dicEntr1 = (dicEnum1.Entry);
+            }
+
+            Assert.Throws<InvalidOperationException>(() =>
+                             {
+                                 strRetKey = (string)dicEnum1.Key;
+                             }
+            );
+
+            //[]We will get a valid enumerator, move to a valid position, then change the underlying HT. Current shouold not throw.
+            //Only calling MoveNext should throw.
+
+            //create the HT
+            hash1 = new Hashtable();
+            for (ii = 0; ii < 100; ii++)
+            {
+                str1 = string.Concat("key_", ii.ToString());
+                hash1.Add(str1, ii);
+            }
+
+            //get the enumerator and move to a valid location
+            dicEnum1 = hash1.GetEnumerator();
+
+            dicEnum1.MoveNext();
+            dicEnum1.MoveNext();
+
+            ii = 87;
+            str1 = string.Concat("key_", ii.ToString());
+
+            //if we are currently pointer at the item that we are going to change then move to the next one
+            if (0 == string.Compare((string)dicEnum1.Key, str1))
+            {
+                dicEnum1.MoveNext();
+            }
+
+            //change the underlying HT
+            hash1[str1] = ii + 50;
+
+            //calling current on the Enumerator should not throw
+            strRetKey = (string)dicEnum1.Key;
+            iExpValue = Convert.ToInt32(dicEnum1.Value);
+
+            Assert.True(hash1.ContainsKey(strRetKey));
+            Assert.True(hash1.ContainsValue(iExpValue));
+
+            strRetKey = (string)dicEnum1.Entry.Key;
+            iExpValue = Convert.ToInt32(dicEnum1.Entry.Value);
+
+            Assert.True(hash1.ContainsKey(strRetKey));
+            Assert.True(hash1.ContainsValue(iExpValue));
+
+            strRetKey = (string)(((DictionaryEntry)(dicEnum1.Current)).Key);
+            iExpValue = Convert.ToInt32(((DictionaryEntry)dicEnum1.Current).Value);
+
+            Assert.True(hash1.ContainsKey(strRetKey));
+            Assert.True(hash1.ContainsValue(iExpValue));
+
+            // calling MoveNExt should throw
+
+            Assert.Throws<InvalidOperationException>(() =>
+                             {
+                                 dicEnum1.MoveNext();
+                             }
+            );
+
+            //[] Try calling clone on the enumerator
+            dicEnum1 = hash1.GetEnumerator();
+            var anotherEnumerator = hash1.GetEnumerator();
+            Assert.False(object.ReferenceEquals(dicEnum1, anotherEnumerator));
+        }
```

Comment:
I didn't see any tests around parallel enumerators, of the sort:

IEnumerator enum1 = hash.GetEnumerator();
IEnumerator enum2 = hash.GetEnumerator();

Assert.True(enum1.MoveNext());

object obj1 = enum1.Current;
object obj2 = null;

while (enum1.MoveNext())
{
    Assert.True(enum2.MoveNext());
    obj2 = enum2.Current;
    Assert.Equals(obj1, obj2);
    obj1 = enum1.Current;
}

Assert.True(enum2.MoveNext());
obj2 = enum2.Current;
Assert.Equal(obj1, obj2);
Assert.False(enum2.MoveNext());


Unless, of course, Hashset doesn't actually support having two outstanding enumerators :)